proposals/0195-dynamic-member-lookup.md

```diff
@@ -0,0 +1,790 @@
+    features, this has become even more common).
+3) it would be *significantly* more invasive in the compiler than this proposal.  While it is
+    unlikely to be as large a scope of impact as the Clang importer in terms of bulk of code, it
+    would require just as many tendrils spread throughout the compiler.
+4)  Taking this approach would set the precedent for all other dynamic languages to get first
+     class language support baked into the Swift compiler, leading to an even greater
+     complexity spiral down the road.
+5) The proposed "first class support" doesn't substantially improve the experience of working
+    with Python over this proposal, so it is all pain and no gain.
+
+Several people have suggested that a "Clang-importer" style Python interoperability approach
+could use the "Type Hints" introduced in
+[PEP 484](https://www.python.org/dev/peps/pep-0484/), which would invalidate that last
+point above.  This approach to progressive typing for Python was prototyped in
+[mypy](http://mypy-lang.org/) and first [shipped in Python
+3.5](https://docs.python.org/3/library/typing.html) in September 2015.
+
+Unfortunately, it isn't reasonable to expect Type Hints to significantly improve the experience
+working with Python in Swift for several reasons, including:
+
+1) PEP 484 states: "It should also be emphasized that **Python will remain a dynamically
+    typed language, and the authors have no desire to ever make type hints mandatory,
+     even by convention.**" (the emphasis is by the authors of PEP 484).  This means we
+     need this proposal or something like AnyObject lookup ... forever.
+2) These annotations are only currently supported on a [provisional
+    basis](https://docs.python.org/3/glossary.html#term-provisional-api), which means that
+    they are "deliberately excluded from .. backwards compatibility guarantees... up to and
+    including removal of the interface".  Because they are subject to change and potentially
+    removal, they haven't gotten wide adoption.
+3) Unlike Objective-C type annotations, these type hints are inherently unsound by design.
+    Further, Python APIs often creep to accepting many more types than a Swift programmer
+    would expect, which means that a type annotation is often "so broad as to be unhelpful" or
+    "too narrow to be correct".
+4) These annotations are not dynamically enforced, and they are specifically designed to
+    "influence" static analysis tools, which means that they are also frequently wrong.  In the
+    context of a narrowly used static analysis tool, a wrong type annotation is merely annoying.
+    In the context of bridging into Swift, being incorrect is a real problem.
+5) The fact that they have only been available in Python **3** (which has a smaller community
+    than Python 2) and have only been shipping for 2 years, means that they haven't been used
+    by many people, and which contributes to their low adoption.
```

Comment:
Does the `/System/Library/Frameworks/Python.framework` on macOS support **Python 3** scripts? If not, that's possibly another reason why Type Hints aren't significant for this proposal.